web/app/solar.js

```diff
@@ -56,72 +74,115 @@ exports.draw = function(canvasSelector, now, grib1, grib2, solarColor, projectio
-                ];
-                var values = positions.map(function(d) {
-                    var n = d[0] + Nx * d[1];
-                    return d3.interpolate(
-                        grib1.data[n],
-                        grib2.data[n])(k);
-                });
-                var val = bilinearInterpolate(
-                    (lonlat[0] - lo1) / dx,
-                    (la1 - lonlat[1]) / dy,
-                    positions[0][0],
-                    positions[2][0],
-                    positions[0][1],
-                    positions[1][1],
-                    values[0],
-                    values[1],
-                    values[2],
-                    values[3]);
-                img.data[((y*(img.width*4)) + (x*4)) + 3] = parseInt(alphas[bucketIndex(val)] * 255);
-            });
+    // Set our domain
+    var NE = projection.invert([realW, 0])
+    var NW = projection.invert([0, 0])
+    var SW = projection.invert([realW, realH])
+    var SE = projection.invert([0, realH])
+    var S = projection.invert([realW / 2, realH])
+    var N = projection.invert([realW / 2, 0])
+    N[1] = 80;
+
+    var minLat = Math.ceil(SE[1]);
+    var maxLat = Math.floor(N[1]);
+    var minLon = Math.ceil(NW[0]);
+    var maxLon = Math.floor(NE[0]);
+    h = maxLat - minLat;
+    w = maxLon - minLon;
+
+    var dt = new Date().getTime();
+    console.log('Draw solar start:' + new Date())
+
+    // Draw initial image (1px 1deg) from grib
+    var imgGrib = ctx.createImageData(w, h);
+    d3.range(minLat, maxLat).forEach(function (y) {
+        d3.range(minLon, maxLon).forEach(function (x) {
+
+            var i = parseInt(x - lo1) / dx;
+            var j = parseInt(la1 - y) / dy;
+            var n = i + Nx * j;
+            var val = d3.interpolate(grib1.data[n], grib2.data[n])(k);
+
+            var pix_x = x - minLon;
+            var pix_y = h - (y - minLat);
+
+            if (levels)
+                imgGrib.data[[((pix_y * (imgGrib.width * 4)) + (pix_x * 4)) + 3]] = parseInt(val / solarScale * 255);
+            else
+                imgGrib.data[[((pix_y * (imgGrib.width * 4)) + (pix_x * 4)) + 3]] = parseInt((0.85 - val / solarScale) * 255);
         });
+    });
+    solarCanvas.attr('height', h);
+    solarCanvas.attr('width', w);
+
+    ctx.clearRect(0, 0, h, w);
+    ctx.putImageData(imgGrib, 0, 0);
+
+    console.log('Extract grib:' + (new Date().getTime() - dt));
+    dt = new Date().getTime();
+
+    // Reproject this image on our projection
+    var sourceData = ctx.getImageData(0, 0, w, h).data,
+        target = ctx.createImageData(realW, realH),
+        targetData = target.data;
 
-        if (x + batchsize <= xrange[xrange.length - 1]) {
-            setTimeout(function() { return batchDrawColumns(x + batchsize, batchsize, callback); }, 25);
-        } else {
-            console.log('Done drawing solar');
-            callback(null);
+    for (var y = 0, i = -1; y < realH; ++y) {
```

Comment:
Why is there a need to iterate in that way? (from i = -1 and increment y in the beginning?)